packages/beacon-node/src/api/impl/beacon/blocks/index.ts

```diff
@@ -381,29 +393,34 @@ export function getBeaconBlockApi({
         if (slot === head.slot) {
           return {
             executionOptimistic: isOptimisticBlock(head),
+            finalized: false,
             data: {root: fromHexString(head.blockRoot)},
           };
         }
 
         if (slot < head.slot && head.slot <= slot + SLOTS_PER_HISTORICAL_ROOT) {
           const state = chain.getHeadState();
+          const rootSlot = slot % SLOTS_PER_HISTORICAL_ROOT;
           return {
             executionOptimistic: isOptimisticBlock(head),
-            data: {root: state.blockRoots.get(slot % SLOTS_PER_HISTORICAL_ROOT)},
+            finalized: computeEpochAtSlot(rootSlot) <= chain.forkChoice.getFinalizedCheckpoint().epoch,
```

Comment:
Epoch calculation should use `slot` instead ie. `computeEpochAtSlot(slot)`

Reply:
Are you sure? We return the root from `slot % SLOTS_PER_HISTORICAL_ROOT` and as per spec
> True if the response references the finalized history of the chain

So shouldn't we always use the slot of the data to determine `finalized=true|false`?

Reply:
> Are you sure? We return the root from `slot % SLOTS_PER_HISTORICAL_ROOT` and as per spec:
> 
> > True if the response references the finalized history of the chain
> 
> So shouldn't we always use the slot of the data to determine `finalized=true|false`?

I think `forkChoice.getFinalizedCheckpoint().epoch` stores the regular (unmodded epoch). But you are comparing it with the epoch calculated from modded slot. 

Say `slot = 1234 + 8192 = 9426` and epoch of finalized checkpoint is `finalizedEpoch = 290`. So epoch of slot is `computeEpochAtSlot(9426) = Math.floor(9426 / 32) = 294`

In your calculation, `rootSlot = 9426 % 8192 = 1234` and `computeEpochAtSlot(rootSlot) = 38`.

I don't think epoch 38 is the correct value to compare with `finalizedEpoch` and will result in `finalized` always be true.

Would love to be corrected if I am wrong @nflaig @tuyennhv

Reply:
I think you are correct, that's even how I had it initially then got confused how SLOTS_PER_HISTORICAL_ROOT works, I have updated the check to use (unmodded) slot to determine epoch